### DISCUSSION TITLE
How to keep Accept Header?

### DISCUSSION THREAD

**Discussion by savage-alex** [apideck-libraries/portman#233] [Q&A] [answered]
Would there be a benefit in configuring the accept header in postman based on the openAPI definition. In the response section it defines the possible media types that can be returned. 

I have a variation test asking for something weird in order to trigger a 406 but sadly I came across an API that rejected an accept header of */* and wanted a specific definition. Obviously this is wrong but it did make me think - normally a consumer is going to ask for a specific media type (99% application/json) so it would be good to specifically ASK for it in the request. It does verify that is what was returned. It might be niche but is it more appropriate to ask for a specific media type via the accept header ?

**Comment by thim81** [comment, marked as answer]
Hi @savage-alex 

Some context - OpenAPI specification: 
> If in is "header" and the name field is "Accept", "Content-Type" or "Authorization", the parameter definition SHALL be ignored.
As defined in [OpenAPI specification](https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.0.3.md#fixed-fields-10), these headers should be ignored as part of the parameter definition.

Portman is using under the hood "[openapi-to-postman](https://github.com/postmanlabs/openapi-to-postman)" to do the initial conversion from OpenAPI to Postman. "openapi-to-postman" follows this OpenAPI recommendation, which results in the conversion dropping all the Accept", "Content-Type" or "Authorization" headers defined in the OpenAPI document, from the generated Postman requests.

But like you said, there are a number of cases, where defining the "Content-Type" or "Accept" or even "Authorization", is a "must" to instruct the REST API about the expected response.

We have provided a [PR](https://github.com/postmanlabs/openapi-to-postman/issues/361) towards "openapi-to-postman" that got merged, which allow you to keep these implicit headers.

So by defining an extended "openapi-to-postman" configuration, you can instruct the conversion to keep the implicit headers in the Postman collection (a list of all available "openapi-to-postman" configuration can be found [here](https://github.com/postmanlabs/openapi-to-postman/blob/develop/OPTIONS.md))

Sample of an extended "openapi-to-postman" configuration, notice the `"keepImplicitHeaders": true`: 
```json
{
  "folderStrategy": "Tags",
  "requestParametersResolution": "Example",
  "exampleParametersResolution": "Example",
  "keepImplicitHeaders": true
}
```

(Check the "[Portman CLI options example](https://github.com/apideck-libraries/portman/tree/main/examples/cli-options)" on how to pass the extended "openapi-to-postman" to Portman. )

This will result in a Postman collection with the "Content-Type" or "Accept" or "Authorization" headers, explicitly defined in OpenAPI.
Plus, you can manipulate these header values with the Portman [overwriteRequestHeaders](https://github.com/apideck-libraries/portman#portman---overwrites-properties) settings to generate variations where the "Accept" type is set to something different.

**Reply by savage-alex** [reply]
Amazing thank you!
